File: julia/SomeParsers.jl
include("Parser.jl")

module SomeParsers

import Monads: (>>)

using ..MonadParser
import ..MonadParser: Parser, mreturn, and_then, |, +, satisfy, optional_s, optional_c, ~, many

export alnum, char, symbol, name, sign, usign, double

alnum::Parser = satisfy(c -> isletter(c) || isdigit(c) || c == '_')

char(x::Char)::Parser = satisfy(c::Char -> c == x)

symbol(c::Char)::Parser = ~char(c)

name(n::String)::Parser = ~and_then(+alnum, s -> s == n ? mreturn(Parser, n) : empty(Parser))

double1::Parser = ~and_then(+satisfy(c -> isdigit(c)), s -> mreturn(Parser, parse(Float64, s)))

digit::Parser = satisfy(c -> isdigit(c))

digits::Parser = many(digit)

sign::Parser = optional_c(char('+') | char('-'))

usign::Parser = optional_c(symbol('+') | symbol('-'))

double::Parser = ~and_then(digits,
    int_part  -> and_then(optional_s(char('.') >> digits),
    frac_part -> and_then(optional_s(and_then((char('e') | char('E')) >> sign,
        exp_sign   -> and_then(+digit,
        exp_digits -> mreturn(Parser, exp_sign * exp_digits)))),
    exp_part  -> length(int_part) > 0 || length(frac_part) > 0 ?
        mreturn(Parser, parse(Float64, int_part *
            (length(frac_part) > 0 ? '.' * frac_part : "") *
            (length(exp_part) > 0 ? 'e' * exp_part : ""))) :
        empty(Parser))))

end # module SomeParsers

File: C#/SomeParsers.cs
namespace MonadParser
{
    class SomeParsers
    {
        public static Parser<char> alnum => Parser<char>.Satisfy(c => Char.IsLetterOrDigit(c) || c == '_');

        public static Parser<char> Char_(char c) => Parser<char>.Satisfy(x => x == c);

        public static Parser<char> Symbol(char c) => Char_(c).Token();

        public static Parser<string> Name(string n) => alnum.Some()
            .FlatMap(s => s.Equals(n) ? Parser<string>.Pure(n) : Parser<string>.Empty()).Token();

        public static Parser<char> digit => Parser<char>.Satisfy(Char.IsDigit);

        public static Parser<string> digits => digit.Many();

        public static Parser<string> sign => Parser<char>.Optional(Char_('+') | Char_('-'));
        public static Parser<string> usign => Parser<char>.Optional(Symbol('+') | Symbol('-'));

        private static string NumberDecimalSeparator = System.Globalization.CultureInfo.CurrentCulture.NumberFormat.NumberDecimalSeparator;

        public static Parser<double> double_ => digits.FlatMap(
            int_part  => Parser<char>.Optional(Char_('.').Skip(digits)).FlatMap(
            frac_part => Parser<char>.Optional((Char_('e') | Char_('E')).Skip(sign).FlatMap(
                exp_sign   => digit.Some().FlatMap(
                exp_digits => Parser<string>.Pure(exp_sign + exp_digits)))).FlatMap(
            exp_part => int_part.Length > 0 || frac_part.Length > 0 ?
                Parser<double>.Pure(double.Parse(int_part +
                    (frac_part.Length > 0 ? NumberDecimalSeparator + frac_part : "") +
                    (exp_part.Length > 0 ? 'e' + exp_part : ""))) :
                Parser<double>.Empty()))).Token();

    }
}
